Replacing unload events handlers to use pagehide

spenserhale opened the pull request:
Chrome has deprecated the 'unload' event (https://developer.chrome.com/blog/deprecating-unload/)

Replacing event handlers to use the 'pagehide' instead based on the suggestion from Chrome's documentation.

Trac ticket: https://core.trac.wordpress.org/ticket/55491

# Commit Message

Administration: Remove deprecated `unload` event handlers and use `pagehide` (and `pageshow`) when appropriate.

Use `pagehide` event instead of `unload` in the following cases:

* For classic editor to release the post lock.
* In Text widget to rebuild editor after dragging widget to new location in classic widgets interface.
* To clear out the `window.name` when navigating away from a post preview.
* To suspend heartbeat, while also using `pageshow` event to resume as if it had been a focused tab in case page restored from bfcache. 

Also:

* Remove obsolete mobile cleanup code in `js/_enqueues/lib/gallery.js` (introduced in [9894]). Do same for `src/js/_enqueues/wp/media/models.js` (introduced in [22872]). See #22552.
* Remove obsolete Firefox-specific workaround in `js/_enqueues/wp/mce-view.js` from [39282]. See #38511.

Fixes #55491.
Props spenserhale, westonruter, adamsilverstein, azaozz, shawfactor, peterwilsoncc, swissspidy.

spenserhale committed:
Replacing unload events handlers to use pagehide

Chrome has deprecated the unload event (https://developer.chrome.com/blog/deprecating-unload/)

Fixes #55491

spenserhale committed:
Replacing unload events handlers to use pagehide

Chrome has deprecated the unload event (https://developer.chrome.com/blog/deprecating-unload/)

Fixes #55491

westonruter commented on src/js/_enqueues/lib/gallery.js:
```diff
@@ -88,7 +88,7 @@ jQuery( function($) {
 });
 
-jQuery(window).on( 'unload', function () { window.tinymce = window.tinyMCE = window.wpgallery = null; } ); // Cleanup.
+jQuery(window).on( 'pagehide', function () { window.tinymce = window.tinyMCE = window.wpgallery = null; } ); // Cleanup.
```
I think we should look into removing this instance entirely. Otherwise, if a tab is restored then it would be broken. And it wouldn't be possible to restore these variables with `pageshow` since they are gone.

westonruter replied:
This change was introduced in r9894

westonruter replied:
Having a hard time finding where this file is even used in core.

westonruter replied:
It is enqueued in two functions: `wp_media_upload_handler()` and `media_upload_gallery()`. This latter function refers to the screen as the "legacy media uploader form".

westonruter replied:
OK, I found out how to get there:

0. Activate the Classic Editor plugin
1. Create a new post
2. Upload _new_ photos in the post and create a gallery from them
3. Save the post
4. With the post ID in hand, go to `/wp-admin/media-upload.php?type=gallery&tab=type&post_id={POST_ID}`

It looks like this:

![image](https://github.com/WordPress/wordpress-develop/assets/134745/ff8b9593-1ccd-4ce8-9630-e0d64cbc083a)

I don't see any other way to get to that screen.

Given this, either we:

1. Replace `unload` with `pagehide` and add a `pageshow` that does the reload behavior below, or
2. We just remove the line of cleanup code altogether. It seems like it was put in place for an iOS Safari bug from over a decade ago, similar to https://github.com/WordPress/wordpress-develop/pull/5056/files#r1332289674, so it's probably obsolete.

@azaozz Since you introduced this line of code and are probably the most familiar with this codebase, what do you think we should do (also considering this is an inaccessible legacy screen)?

westonruter replied:
See also how @spenserhale removed similar code in `src/js/_enqueues/wp/mce-view.js`: https://github.com/WordPress/wordpress-develop/pull/5056/files#r1332287807

westonruter replied:
I'm just going to go ahead and remove this for consistency with `src/js/_enqueues/wp/mce-view.js`.

westonruter commented on src/js/_enqueues/wp/mce-view.js:
```diff
@@ -663,7 +663,7 @@
 				}
 
-				$( iframeWin ).on( 'load', resize ).on( 'unload', reload );
+				$( iframeWin ).on( 'load', resize ).on( 'pagehide', reload );
```
Introduced in r39282 to fix Core-38511 in 2016. It is to work-around a Firefox-specific bug. We should validate whether it is still an issue in Firefox. If not, we can remove this code.

westonruter commented on src/js/_enqueues/wp/media/models.js:
```diff
@@ -239,6 +239,6 @@ media.query = function( props ) {
 
 // Clean up. Prevents mobile browsers caching.
-$(window).on('unload', function(){
+$(window).on('pagehide', function(){
```
Origin is https://github.com/WordPress/wordpress-develop/commit/d2d01ad797e7894b6ea6d367c973f0c3d670003b to fix Core-22552: "Empty wp on unload to work around caching in iOS Safari."

This commit is 11 years old. Most likely this is no longer an issue in Safari. If so, we can most likely delete this entire event handler.

westonruter commented on src/js/_enqueues/wp/widgets/text.js:
```diff
@@ -291,7 +291,7 @@ wp.textWidgets = ( function( $ ) {
 
 					// When a widget is moved in the DOM the dynamically-created TinyMCE iframe will be destroyed and has to be re-built.
-					$( editor.getWin() ).on( 'unload', function() {
+					$( editor.getWin() ).on( 'pagehide', function() {
```
The use of the `unload` event may not be a problem here for bfcache since it is added to the iframe and not the window as a whole. Still, if `pagehide` works the same we could still use it.

westonruter replied:
Using `unload` is indeed a problem here. It disabled bfcache:

> ![image](https://github.com/WordPress/wordpress-develop/assets/134745/7a406fce-5e6b-4316-9e3d-7c82f4a5246c)

westonruter replied:
Removing the handler altogether results in a breakage:

[Screen recording 2023-09-20 17.09.09.webm](https://github.com/WordPress/wordpress-develop/assets/134745/24344ca7-0191-4618-9dde-1ac7edf36f51)

Nevertheless, using `pagehide` instead of `unload` does work:

[Screen recording 2023-09-20 17.09.40.webm](https://github.com/WordPress/wordpress-develop/assets/134745/3bd5e06b-4df2-4295-a159-a45ecb473260)

westonruter replied:
Nevertheless, bfcache remains disabled since TinyMCE is using `unload`. However, this doesn't seem like something we should remove per https://core.trac.wordpress.org/ticket/55491#comment:27

westonruter commented:
The heartbeat instance we also need to account for:

https://github.com/WordPress/wordpress-develop/blob/81423c3ea63531c8c9ff7591295825eabd7a5b99/src/js/_enqueues/wp/heartbeat.js#L226

westonruter commented:
Another case to investigate:

https://github.com/WordPress/wordpress-develop/blob/81423c3ea63531c8c9ff7591295825eabd7a5b99/src/js/_enqueues/admin/post.js#L514

westonruter committed:
Add pageshow handler for heartbeat

spenserhale committed:
MCE-View: Removing pagehide event listener

reload function was to fix #38511, but Firefox issue is no longer reproducible

For Ticket #55491

westonruter committed:
Remove obsolete unload handler from Sizzle library bundled with TinyMCE

spenserhale committed:
Media/models.js Removing pagehide event listener

was to fix #22552, but the iOS issue is no longer reproducible

For Ticket #55491

westonruter committed:
Make sure post lock is re-established when editor is restored from bfcache

westonruter committed:
Merge branch 'trunk' into 55491-replace-unload-events

westonruter committed:
Reload page upon persisted pageshow on page enqueueing gallery.js

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into 55491-replace-unload-events

westonruter committed:
Revert patch to tinymce

westonruter commented on src/js/_enqueues/wp/mce-view.js:
```diff
@@ -643,16 +643,6 @@
 					}, 3000 );
 				}
 
-				function reload() {
```
Rationale for removal: https://core.trac.wordpress.org/ticket/55491#comment:15

westonruter replied:
Introduced in r39282 to fix Core-38511 in 2016. It is to work-around a Firefox-specific bug. We should validate whether it is still an issue in Firefox. If not, we can remove this code.

westonruter replied:
I tried reproducing the issue in Firefox according to the description of Core-38511 and I was unable.

westonruter replied:
I've reported this to the 7-year old ticket: https://core.trac.wordpress.org/ticket/38511#comment:27

westonruter replied:
@azaozz also please weigh in on whether we can remove this code.

adamsilverstein replied:
Given the age of this bug fix, it should be fine to remove. If we discover that this still creates an issue for Firefox users, we can always restore the fix.

azaozz replied:
Yep, super old browser specific fix. LGTM to remove too.

westonruter commented on src/js/_enqueues/wp/media/models.js:
```diff
@@ -237,8 +237,3 @@ media.query = function( props ) {
 	});
 };
-
-// Clean up. Prevents mobile browsers caching.
-$(window).on('unload', function(){
-	window.wp = null;
-});
```
Rationale for removal: https://core.trac.wordpress.org/ticket/55491#comment:17

However, if this depends on the server sending `Cache-Control: no-store`, then I'd like to further verify this. 

@spenserhale What happens if `no-store` is _not_ sent? Is there an issue then? I'd like to find a way to not have to send `no-store` in the admin so that we can benefit from bfcache, but if this depends on `no-store` then that would block enabling bfcache.

westonruter replied:
Origin is https://github.com/WordPress/wordpress-develop/commit/d2d01ad797e7894b6ea6d367c973f0c3d670003b to fix Core-22552: "Empty wp on unload to work around caching in iOS Safari."

This commit is 11 years old. Most likely this is no longer an issue in Safari. If so, we can most likely delete this entire event handler.

adamsilverstein replied:
+1 - fine to delete this super old bug fix/browser specific workaround. It is unlikely that the bug still exists, if it does we can always reintroduce a fix.

azaozz replied:
Yep, same as above. Super old browser specific fix that seems okay to remove

westonruter commented on src/wp-includes/functions.php:
```diff
@@ -7665,7 +7665,7 @@ function wp_post_preview_js() {
 
 		if ( window.addEventListener ) {
-			window.addEventListener( 'unload', function() { window.name = ''; }, false );
+			window.addEventListener( 'pagehide', function() { window.name = ''; } );
```
I was confused why the `PageTransitionEvent` being passed to this event handler has its `persisted` property set to `false`, both for the `pagehide` and the `pageshow` events. In looking at DevTools, I see:

> ![image](https://github.com/WordPress/wordpress-develop/assets/134745/e26f70b1-7665-4ed1-a5a9-9ad8fd211a5e)

So it is explained that the fact that `window.open()` was used to open the preview is another way that bfcache is disabled. 

I think this is all good. I'm seeing that the `window.name` is being set to `wp-preview-...` when the tab is opened when clicking preview. When I navigate the preview window to another page, the `window.name` is set to an empty string. When I then re-preview a post I am opened in a new tab rather than the previous preview tab (which is now no longer a preview). If in that preview tab I go back to the previewed post, and then hit preview again, then it is that tab that gets updated with the latest preview content.

So I think we're good here.

westonruter commented on src/js/_enqueues/wp/heartbeat.js:
```diff
@@ -223,16 +223,33 @@
 				}
 			});
 
+			$(window).on(
+				'pageshow.wp-heartbeat',
+				/**
+				 * Handles pageshow event.
+				 *
+				 * @param {jQuery.Event} event
+				 * @param {PageTransitionEvent} event.originalEvent
+				 */
+				function ( event ) {
+					if ( event.originalEvent.persisted ) {
+
+						// Continue connecting.
+						resume();
+					}
+				}
+			);
```
I'm not seeing this code firing.

westonruter replied:
OK, I am seeing it fire. Apparently I have a browser extension which is disabling bfcache, but if I try in an incognito window then `event.originalEvent.persisted` is `true`.

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into 55491-replace-unload-events

westonruter commented on src/js/_enqueues/admin/post.js:
```diff
@@ -511,7 +511,7 @@ jQuery( function($) {
 			return __( 'The changes you made will be lost if you navigate away from this page.' );
 		}
-	}).on( 'unload.edit-post', function( event ) {
+	}).on( 'pagehide.edit-post', function( event ) {
```
I don't see a single reference to the jQuery-namespaced `unload.edit-post` event name in WPdirectory, so I don't see any back-compat concerns here.

westonruter replied:
Note that post lock is restored upon `pageshow` via heartbeat. See https://github.com/WordPress/wordpress-develop/pull/5056#discussion_r1343245997

westonruter commented on src/js/_enqueues/wp/heartbeat.js:
```diff
@@ -223,16 +223,33 @@
 			}
 
-			$(window).on( 'unload.wp-heartbeat', function() {
+			$(window).on( 'pagehide.wp-heartbeat', function() {
```
As with `pagehide.edit-post`, I don't see a single reference to the jQuery-namespaced `unload.wp-heartbeat` event name in WPdirectory, so I don't see any back-compat concerns here.

westonruter replied:
As with `pagehide.edit-post`, I don't see a single reference to the jQuery-namespaced `unload.wp-heartbeat` event name in WPdirectory, so I don't see any back-compat concerns here.

westonruter commented on src/js/_enqueues/admin/post.js:
```diff
@@ -557,7 +557,20 @@ jQuery( function($) {
+		 */
+		function ( event ) {
+			if ( event.originalEvent.persisted ) {
+				window.location.reload();
```
In practice this code will never fire because this code is loaded on the Classic Editor and TinyMCE disables bfcache due to an `unload` event handler.

westonruter replied:
Nevertheless, _if_ that offending TinyMCE code is commented-out _and_ `Cache-Control: no-store` is disabled (e.g. via [plugin](https://gist.github.com/westonruter/8c19d87a80a36e8f24db910750162628)), then the experience is not great when you leave with a dirty editor state: you get the "Are you sure?" dialog when you leave and then when you go back, to get it again when it attempts to reload: 

[Screen recording 2023-10-02 16.03.53.webm](https://github.com/WordPress/wordpress-develop/assets/134745/9c8bc069-38e3-4c5e-8d6f-ead0112773ae)

westonruter replied:
In the end, this code is not needed because upon `pageshow`, heartbeat will `resume` and it will start sending its `wp-refresh-post-lock` requests.

westonruter replied:
So this `pageshow` event is removed in 329ebfbb2932dbddb69a66d71231131c60d4b8d3, and then heartbeat then connects immediately upon `pageshow` if restoring from bfcache in 00772ee2907d2b736c1fbbb36871f641fe4bbe7d. 

Much better experience:

[Screen recording 2023-10-02 16.19.16.webm](https://github.com/WordPress/wordpress-develop/assets/134745/b9c0b1c8-6332-4677-b471-a3e9dd1042a6)

westonruter committed:
Remove unnecessary pageshow.edit-post event handler to reload page when restoring from bfcache

westonruter committed:
Connect heartbeat immediately when restoring from bfcache

westonruter committed:
Use focused() function instead of resume and connectNow, since it does both of those

westonruter committed:
Remove unused jQuery reference

westonruter committed:
Remove (likely) obsolete cleanup code

westonruter committed:
Also remove pageshow event after removal of pagehide

westonruter commented on src/js/_enqueues/wp/heartbeat.js:
```diff
@@ -223,16 +223,35 @@
+				 * @param {PageTransitionEvent} event.originalEvent
+				 */
+				function ( event ) {
+					if ( event.originalEvent.persisted ) {
+						/*
+						 * When page navigation is stored via bfcache (Back/Forward Cache), consider this the same as
+						 * if the user had just switched to the tab since the behavior is similar.
+						 */
+						focused();
+					}
```
See outdated conversation: https://github.com/WordPress/wordpress-develop/pull/5056#discussion_r1343238144

westonruter commented on src/js/_enqueues/lib/gallery.js:
```diff
@@ -88,8 +88,6 @@ jQuery( function($) {
 	}
 });
 
-jQuery(window).on( 'unload', function () { window.tinymce = window.tinyMCE = window.wpgallery = null; } ); // Cleanup.
-
```
Removal rationale: https://github.com/WordPress/wordpress-develop/pull/5056/files#r1332287807

See also https://github.com/WordPress/wordpress-develop/pull/5056#discussion_r1304429877

azaozz commented:
The changes from `unload` to `pagehide` seem pretty straightforward. Also the removal of super old browser specific bugfixes. On the other hand this touches an area where bugs may be harder to figure out, but thinking that if something goes wrong it would be detected and fixed during RC.

westonruter commented:
Committed in r56809.
